pep-0696.rst

```diff
@@ -0,0 +1,421 @@
+       AllTheDefaults[int, complex, str, int] ==
+       AllTheDefaults[int, complex, str, int, bool]
+   )  # All valid
+
+This cannot be enforced at runtime for functions, for now, but in the
+future, this might be possible (see `Interaction with PEP
+695 <#interaction-with-pep-695>`__).
+
+``Generic`` ``TypeAlias``\ es
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``Generic`` ``TypeAlias``\ es should be able to be further subscripted
+following normal subscription rules. If a ``TypeVar`` has a default
+that hasn't been overridden it should be treated like it was
+substituted into the ``TypeAlias``. However, it can be specialised
+further down the line.
+
+.. code:: py
+
+   class SomethingWithNoDefaults(Generic[T, T2]): ...
+
+   MyAlias: TypeAlias = SomethingWithNoDefaults[int, DefaultStrT]  # valid
+   reveal_type(MyAlias)        # type is SomethingWithNoDefaults[int, str]
+   reveal_type(MyAlias[bool])  # type is SomethingWithNoDefaults[int, bool]
+
+   MyAlias[bool, int]  # Invalid: too many arguments passed to MyAlias
+
+Subclassing
+~~~~~~~~~~~
+
+Subclasses of ``Generic``\ s with ``TypeVar``\ s that have defaults
+behave similarly to ``Generic`` ``TypeAlias``\ es.
+
+.. code:: py
+
+   class SubclassMe(Generic[T, DefaultStrT]): ...
+
+   class Bar(SubclassMe[int, DefaultStrT]): ...
+   reveal_type(Bar)        # type is Bar[str]
+   reveal_type(Bar[bool])  # type is Bar[bool]
+
+   class Foo(SubclassMe[int]): ...
+
+   reveal_type(Spam)  # type is <subclass of SubclassMe[int, int]>
+   Foo[str]  # Invalid: Foo cannot be further subscripted
+
+   class Baz(Generic[DefaultIntT, DefaultStrT]): ...
+
+   class Spam(Baz): ...
+   reveal_type(Spam)  # type is <subclass of Baz[int, str]>
+
+Using bound and default
+~~~~~~~~~~~~~~~~~~~~~~~
+
+If both ``bound`` and ``default`` are passed ``default`` must be a
+subtype of ``bound`` otherwise the type checker should generate an
+error.
+
+.. code:: py
+
+   TypeVar("Ok", bound=float, default=int)     # Valid
+   TypeVar("Invalid", bound=str, default=int)  # Invalid: the bound and default are incompatible
+
+Constraints
+~~~~~~~~~~~
+
+For constrained ``TypeVar``\ s, the default needs to be one of the
+constraints. A type checker should generate an error even if it is a
+subtype of one of the constraints.
+
+.. code:: py
+
+   TypeVar("Ok", float, str, default=float)     # Valid
+   TypeVar("Invalid", float, str, default=int)  # Invalid: expected one of float or str got int
+
+Function Defaults
+~~~~~~~~~~~~~~~~~
+
+The ``TypeVar``\ 's default should also be compatible with the
+parameter's runtime default if present, if they aren't a type checker
+should generate an error.
+
+.. code:: py
+
+   def foo(x: DefaultIntT = 12345): ...  # Valid
+   def foo(x: DefaultIntT = "bar"): ...  # Invalid: expected an int for default, got str
+
+The defaults are erased when called so, they act like they have no default.
+
+.. code:: py
+
+   def baz(y: DefaultIntT) -> DefaultIntT: ...
```

Comment:
This is a weird example, because the default will never be used. A type checker could justifiably show a warning for this. I'm not sure the example is useful.